src/main/java/ch/njol/skript/lang/function/FunctionReference.java

```diff
@@ -277,28 +282,65 @@ public boolean resetReturnValue() {
 		// Prepare parameter values for calling
 		Object[][] params = new Object[singleListParam ? 1 : parameters.length][];
 		if (singleListParam && parameters.length > 1) { // All parameters to one list
-			List<Object> l = new ArrayList<>();
+			params[0] = evaluateSingleListParameter(parameters, event, function.getParameter(0).keyed);
+		} else { // Use parameters in normal way
+			for (int i = 0; i < parameters.length; i++)
+				params[i] = evaluateParameter(parameters[i], event, function.getParameter(i).keyed);
+		}
+
+		// Execute the function
+		return function.execute(params);
+	}
+
+	private Object[] evaluateSingleListParameter(Expression<?>[] parameters, Event event, boolean keyed) {
+		if (!keyed) {
+			List<Object> list = new ArrayList<>();
 			for (Expression<?> parameter : parameters)
-				l.addAll(Arrays.asList(parameter.getArray(event)));
-			params[0] = l.toArray();
+				list.addAll(Arrays.asList(evaluateParameter(parameter, event, false)));
+			return list.toArray();
+		}
+
+		List<Object> values = new ArrayList<>();
+		Set<String> keys = new LinkedHashSet<>();
+		int keyIndex = 1;
+		for (Expression<?> parameter : parameters) {
+			Object[] valuesArray = parameter.getArray(event);
+			String[] keysArray = parameter instanceof KeyProviderExpression<?> keyProvider && keyProvider.canReturnKeys() && keyProvider.areKeysRecommended()
+				? keyProvider.getArrayKeys(event)
+				: null;
 
 			// Don't allow mutating across function boundary; same hack is applied to variables
-			for (int i = 0; i < params[0].length; i++) {
-				params[0][i] = Classes.clone(params[0][i]);
+			for (Object value : valuesArray)
+				values.add(Classes.clone(value));
+
+			if (keysArray != null) {
+				keys.addAll(Arrays.asList(keysArray));
+				continue;
 			}
-		} else { // Use parameters in normal way
-			for (int i = 0; i < parameters.length; i++) {
-				Object[] array = parameters[i].getArray(event);
-				params[i] = Arrays.copyOf(array, array.length);
-				// Don't allow mutating across function boundary; same hack is applied to variables
-				for (int j = 0; j < params[i].length; j++) {
-					params[i][j] = Classes.clone(params[i][j]);
-				}
+
+			for (int i = 0; i < valuesArray.length; i++) {
+				while (keys.contains(String.valueOf(keyIndex)))
+					keyIndex++;
+				keys.add(String.valueOf(keyIndex));
```

Comment:
should probably increment to avoid an extra contains check
```suggestion
				keys.add(String.valueOf(keyIndex++));
```